pkg/clusteragent/autoscaling/workload/loadstore/store.go

```diff
@@ -0,0 +1,115 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver
+
+package loadstore
+
+import (
+	"math"
+	"strings"
+
+	"github.com/DataDog/agent-payload/v5/gogen"
+)
+
+// StoreInfo represents the store information, like memory usage and entity count.
+type StoreInfo struct {
+	TotalMemoryUsage       uint64
+	TotalEntityCount       uint64
+	currentTime            Timestamp
+	EntityCountByMetric    map[string]uint64
+	EntityCountByNamespace map[string]uint64
+}
+
+// Store is an interface for in-memory storage of entities and their load metric values.
+type Store interface {
+	// SetEntitiesValues sets the values for the given map
+	SetEntitiesValues(entities map[*Entity]*EntityValue)
+
+	// GetStoreInfo returns the store information.
+	GetStoreInfo() StoreInfo
+
+	// GetEntitiesByNamespace to get all entities and values by namespace
+	GetEntitiesByNamespace(namespace string) map[*Entity]*EntityValue
+
+	// GetEntitiesByMetricName to get all entities and values by load metric name
+	GetEntitiesByMetricName(metricName string) map[*Entity]*EntityValue
+
+	// GetAllMetricNamesWithCount to get all metric names and corresponding entity count
+	GetAllMetricNamesWithCount() map[string]int64
+
+	// GetAllNamespaceNamesWithCount to get all namespace names and corresponding entity count
+	GetAllNamespaceNamesWithCount() map[string]int64
+
+	// GetEntityByHashKey to get entity and latest value by hash key
+	GetEntityByHashKey(hash uint64) (*Entity, *EntityValue)
+
+	//DeleteEntityByHashKey to delete entity by hash key
+	DeleteEntityByHashKey(hash uint64)
+}
+
+// createEntitiesFromPayload is a helper function used for creating entities from the metric payload.
+func createEntitiesFromPayload(payload *gogen.MetricPayload) map[*Entity]*EntityValue {
+	entities := make(map[*Entity]*EntityValue)
+	splitTag := func(tag string) (key string, value string) {
+		split := strings.SplitN(tag, ":", 2)
```

Comment:
nit: We might want to use https://pkg.go.dev/strings#Index to optimize this further.